File: slogger/src/main/java/io/github/heartinfei/slogger/plan/BasePlan.java
package io.github.heartinfei.slogger.plan;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import io.github.heartinfei.slogger.Configuration;

/**
 * 简介：
 *
 * @author 王强 on 2017/11/23 249346528@qq.com
 */
public abstract class BasePlan {
    /**
     * Android fragment 内置的Log工具类限制输出长度不能超过4k，所以要打印的log信息如果超过4k需要做分块打印处理
     */
    private static final int CHUNK_SIZE = 4000;

    /**
     * 输出Log信息
     *
     * @param c   配置信息
     * @param msg log内容
     */
    protected void logInfo(@NonNull Configuration c, @Nullable List<String> msg) {
    }

    /**
     * 输出错误信息
     *
     * @param c
     * @param msg
     */
    protected void logErro(@NonNull Configuration c, @Nullable List<String> msg) {
    }

    protected void i(Configuration conf, String... msgs) {
        String header = generateLogHeader(conf);
        for (String msg : msgs) {
            logInfo(conf, buildMessage(header, msg, conf));
        }
    }

    protected void e(Configuration conf, Throwable... msgs) {
        for (Throwable t : msgs) {
            logErro(conf, buildMessage(null, getThrowableInfo(t), conf));
        }
    }

    private List<String> buildMessage(String header, @Nullable String msg, @NonNull Configuration config) {
        if (msg == null) {
            msg = "";
        }
        if (header == null) {
            header = "";
        }
        List<String> result = new LinkedList<>();
        //要打印的log 日志的总长度
        int len = header.length() + msg.length();
        if (len < CHUNK_SIZE) {
            result.add(header + msg);
        } else {
            int lastIndex = 0;
            for (int i = CHUNK_SIZE - header.length(); i < len; i += CHUNK_SIZE) {
                result.add(header + msg.substring(lastIndex, i));
                lastIndex += i;
            }
        }
        return result;
    }

    /**
     * 生成Log的头信息
     *
     * @param config 配置信息
     * @return 头信息
     */
    private String generateLogHeader(Configuration config) {
        StringBuilder msgBuilder = new StringBuilder();
        if (config.isShowThreadInfo()) {
            msgBuilder.append(getThreadInfo());
        }
        //当前App的堆栈信息
        List<String> trackInfos =
                getTrackInfo(extractCurrentAppStackTrack(config.getPkgName()), config.getTrackInfoDeep());
        if (config.isPrintLineNo()) {
            msgBuilder.append(trackInfos.remove(0));
        }
        if (config.isPrintTrackInfo()) {
            for (String trackInfo : trackInfos) {
                msgBuilder.append(trackInfo);
            }
        }
        return msgBuilder.toString();
    }

    /**
     * 生成当前App调用关系的堆栈信息
     *
     * @param pkgName 当前App的包名
     * @return 堆栈信息
     */
    private List<StackTraceElement> extractCurrentAppStackTrack(String pkgName) {
        List<StackTraceElement> result = new ArrayList<>();
        try {
            for (StackTraceElement traceElement : Thread.currentThread().getStackTrace()) {
                if (traceElement.getClassName().contains(pkgName)) {
                    result.add(traceElement);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return result;
    }

    /**
     * 生成堆栈信息
     *
     * @param traceElements
     * @param deep
     * @return
     */
    private List<String> getTrackInfo(@NonNull List<StackTraceElement> traceElements, int deep) {
        List<String> result = new ArrayList<>();
        String offset = "";
        deep = deep >= traceElements.size() ? traceElements.size() : 0;
        for (int i = 0; i < deep; i++) {
            StackTraceElement traceElement = traceElements.get(i);
            result.add(offset + traceElement.toString() + "\n");
            offset += "↳";
        }
        return result;
    }

    /**
     * 获取当前线程的信息
     *
     * @return 当前线程的名称
     */
    private String getThreadInfo() {
        return "Thread:" + Thread.currentThread().getName() + "#";
    }

    /**
     * 提取异常信息
     *
     * @param t 异常
     * @return 信息
     */
    private String getThrowableInfo(Throwable t) {
        StringWriter sw = new StringWriter(1024);
        PrintWriter pw = new PrintWriter(sw, false);
        t.printStackTrace(pw);
        pw.flush();
        return sw.toString();
    }
}

File: slogger/src/main/java/io/github/heartinfei/slogger/plan/BasePlan.kt
package io.github.heartinfei.slogger.plan

import android.os.Build
import android.util.Log
import io.github.heartinfei.slogger.LogPrinterProxy
import io.github.heartinfei.slogger.S
import io.github.heartinfei.slogger.SConfiguration
import java.util.regex.Pattern
import kotlin.math.max

/**
 * @author Rango on 2019-05-29 249346528@qq.com
 */
abstract class BasePlan {
    protected open val stackIgnoreFilter = mutableListOf<String>(
            S::class.java.name,
            S.Companion::class.java.name,
            BasePlan::class.java.name,
            LogPrinterProxy::class.java.name
    )

    @get:JvmSynthetic
    internal val explicitTag = ThreadLocal<String>()

    internal open val mTag: String?
        @JvmSynthetic
        get() {
            val tag = explicitTag.get()
            if (tag != null) {
                explicitTag.remove()
            }
            return tag
        }

    /**Add a stack filter.*/
    fun addStackIgnoreFilter(filter: String) {
        stackIgnoreFilter.add(filter)
    }

    /**
     * Return true should be logged.
     */
    protected open fun isLoggable(tag: String?, priority: Int): Boolean = true

    /**
     * Write a echoLog message to its destination. Called for all level-specific methods by default.
     *
     * @param priority Log level. See [Log] for constants.
     * @param tag Explicit or inferred mTag. May be `null`.
     * @param message Formatted echoLog message. May be `null`.
     */
    protected abstract fun echoLog(priority: Int, tag: String?, message: String)

    internal fun assembleAndEchoLog(config: SConfiguration?, priority: Int, message: String?, vararg args: Any?) {
        var tag = if (config?.tag?.isNotEmpty() == true) {
            config.tag
        } else {
            mTag
        }
        if (config?.printThreadInfo == true) {
            tag = tag.plus("Thread#${Thread.currentThread().name}")
        }
        // Tag length limit was removed in API 24.
        if (tag!!.length > MAX_TAG_LENGTH && Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            tag = tag.substring(0, MAX_TAG_LENGTH)
        }

        if (!isLoggable(tag, priority)) {
            return
        }
        config?.apply {
            val header = assembleContentHeader(this)
            val content = assembleLogBody(header, message, *args)
            echoLog(priority, tag, content)
        }
    }

    private fun getStackInfo(config: SConfiguration): List<StackTraceElement>? {
        return if (config.printTrackInfo) {
            Thread.currentThread().stackTrace
                    .filter { element ->
                        val content = element.className
                        val defaultFilterResult = content !in stackIgnoreFilter
                        val configFilter = config.trackFilter.orEmpty()
                        val configFilterResult = content.contains(configFilter)
                        return@filter defaultFilterResult && configFilterResult
                    }
                    .toList()
                    .takeLast(max(0, config.trackDeep))
                    .reversed()
        } else null
    }

    private fun assembleContentHeader(config: SConfiguration): String {
        val headerBuilder = StringBuffer()
        getStackInfo(config)?.let {
            headerBuilder.append(" \n")
            var offset = ""
            var indicator = ""
            for (element in it) {
                headerBuilder.append("$offset$indicator$element\n")
                indicator = "↳"
                offset += " "
            }
            val len = headerBuilder.length
            if (len > 1 && '\n' == (headerBuilder[len - 1])) {
                headerBuilder.deleteCharAt(headerBuilder.length - 1).append("\n")
            } else {
                headerBuilder.append("\n")
            }
        }
        return headerBuilder.toString()
    }

    /**
     * Get current thread name.
     */
    private fun getThreadInfo(config: SConfiguration): String {
        return if (config.printThreadInfo) Thread.currentThread().name else ""
    }


    private fun assembleLogBody(header: String, msg: String?, vararg args: Any?): String {
        var content = header + msg
        for (arg in args) {
            content += arg
        }
        return content
    }

    companion object {
        const val MAX_LOG_LENGTH = 4000
        const val MAX_TAG_LENGTH = 23
        //Extract default log withTag when withTag is null.
        val ANONYMOUS_CLASS = Pattern.compile("(\\$\\d+)+$")
    }

}
